gprofiler/profilers/profiler_base.py

```diff
@@ -119,41 +157,162 @@ def _profiling_error_stack(
-                    )
-                except Exception as e:
-                    logger.exception(f"{self.__class__.__name__}: failed to profile process {pid} ({comm})")
-                    result = ProfileData(
-                        self._profiling_error_stack("error", f"exception {type(e).__name__}", comm), None, None
-                    )
-
-                results[pid] = result
+            return self._wait_for_profiles(futures)
+
+
+class SpawningProcessProfilerBase(ProcessProfilerBase):
+    """
+    Enhances ProcessProfilerBase with tracking of newly spawned processes.
+    """
+
+    _SCHED_THREAD_INTERVAL = 0.1
+    _BACKOFF_INIT = 0.1
+    # so we wait up to 1.5 seconds
+    _BACKOFF_MAX = 0.8
+
+    def __init__(
+        self,
+        frequency: int,
+        duration: int,
+        stop_event: Optional[Event],
+        storage_dir: str,
+        profile_spawned_processes: bool,
+    ):
+        super().__init__(frequency, duration, stop_event, storage_dir)
+        self._profile_spawned_processes = profile_spawned_processes
+        self._submit_lock = Lock()
+        self._threads: Optional[ThreadPoolExecutor] = None
+        self._start_ts: Optional[float] = None
+        self._preexisting_pids: Optional[List[int]] = None
+        self._enabled_proc_events_spawning = False
+        self._futures: Dict[Future, Tuple[int, str]] = {}
+        self._sched = sched.scheduler()
+        self._sched_stop = False
+        self._sched_thread = Thread(target=self._sched_thread_run)
+
+    def _should_profile_process(self, process: Process) -> bool:
+        raise NotImplementedError
+
+    def _notify_selected_processes(self, processes: List[Process]) -> None:
+        # now we start watching for new processes.
+        self._start_profiling_spawning(processes)
+
+    @property
+    def _is_profiling_spawning(self) -> bool:
+        return self._threads is not None
+
+    def _start_profiling_spawning(self, processes: List[Process]) -> None:
+        with self._submit_lock:
+            self._start_ts = time.monotonic()
+            # arbitrary high number of threads to make sure we can run profiling of many
+            # processes concurrently
+            self._threads = ThreadPoolExecutor(max_workers=999)
+            # TODO: add proc_events exit action to remove these
+            self._preexisting_pids = [p.pid for p in processes]
+
+    def _stop_profiling_spawning(self) -> None:
+        with self._submit_lock:
+            self._start_ts = None
+            assert self._threads is not None
+            threads = self._threads
+            # delete it before blocking on the exit of all threads (to ensure no new work
+            # is added)
+            self._threads = None
+            threads.shutdown()  # waits (although - all are done by now)
+            self._preexisting_pids = None
+
+    def _proc_exec_callback(self, tid: int, pid: int) -> None:
+        with contextlib.suppress(NoSuchProcess):
```

Comment:
Should check here if we're not stopping, otherwise we might still block forever in `self._sched.run()`
```suggestion
        if not self._is_profiling_spawning:
            return

        with contextlib.suppress(NoSuchProcess):
```